Added support to line granularity

jose opened the pull request:
Hi @andrefreitas,

Here is a first attempt to address #3.  Please let me know what am I missing or what might be incorrect.

-- Best, Jose

jose committed:
Added support to line granularity (work in progress).

andrefreitas commented on schwa/analysis/repository_analytics.py:
```diff
@@ -252,6 +251,7 @@ def compute_defect_probability(self):
         self.defect_prob = self.defect_probability()
         for class_analytics in self.classes_analytics.values():
             class_analytics.compute_defect_probability()
+        # TODO should we also compute_defect_probability of lines?
```
Yes, the leafs of File analytics can also be Lines and not only Classes.

andrefreitas commented on schwa/analysis/repository_analytics.py:
```diff
@@ -307,4 +308,22 @@ def to_dict(self, name):
+    def to_dict(self, name):
+        metrics_dict = super().to_dict()
+        metrics_dict["type"] = "line"
+        metrics_dict["name"] = name
```
instead of name we need to know the line number. Eventually also the path of the file to be more explicit

jose replied:
Ideally, I would like to represent three different types of lines:
- Lines of code that can be in a file but not in a class or method (e.g., import statement).
- Lines of code that can be in a file and in a class but not in a method (e.g., a class' field).
- Lines of code that can be in a file and in a class and in a method.

I believe commit [c023858](https://github.com/andrefreitas/schwa/pull/5/commits/c023858b4aad707681900b6a81a7f08dd57e1c46) implements that, but I'm still not sure whether the `diff` method in the `java_parser.py` file is correctly implemented.  Could you please check that?

andrefreitas commented on schwa/extraction/git_extractor.py:
```diff
@@ -145,22 +147,27 @@ def extract_commit(self, hexsha):
+                        diffs_list.append(DiffClass(file_name=blob.path, class_b=c, added=True))
+                    for c, m in methods_set:
+                        diffs_list.append(DiffMethod(file_name=blob.path, class_name=c, method_b=m, added=True))
+            if self.line_granularity:
```
Line granularity can be independent from Parsing capabilities of the source code. E.g. Although we cannot parse Python, we can compute what lines were changed, added and removed.

jose replied:
Good point.  Fixed.

andrefreitas reviewed (commented):
Thanks for this contribution @jose and for the time you invested in learning how to change the code 🥇  I left some comments since the line granularity can be independent of the Java parsing capabilities.

I need to make this project run in my computer. I was installing today and one dependency is broken (GitPython) because the doesn't lock transitive dependencies, so an newer release made it to break: https://github.com/gitpython-developers/GitPython/issues/983 I am dockerizing it also.

jose committed:
Line granularity can be independent from Parsing capabilities of the source code.  For example, although schwa does not support Python, it can compute what lines were changed, added and removed.

- A line of code can be in a file but not in a class or method (e.g., import statement).
- A line of code can be in a file and in a class but not in a method (e.g., a class' field).
- A line of code can be in a file and in a class and in a method.

jose committed:
Added missing lists.

jose committed:
Added argument to control granularity level.

jose committed:
Refactored 'Components' and use objects of components rather than strings of components.

jose committed:
Replaced local plyj with [javalang](https://github.com/c2nes/javalang) which seems to be more complete and up-to-date.

jose committed:
Added support for methods with same name but different signatures.

jose committed:
Not just "normal" classes but also interfaces.

jose committed:
Fixed parent component.

jose committed:
Increased warning level.

jose committed:
Simplified analytics by moving common code to super().

jose committed:
The 'repository' does not require a type and metrics field.

jose committed:
Fix, do not include lines of code of subclasses, which in practice belong to the inner/sub-class.

jose committed:
Added support to functions in functions.

jose committed:
Improved 'hash' function.

jose committed:
Make parent argument mandatory.

jose committed:
Keep a component object and not a raw type.

jose committed:
Create 'File' objects for each 'DiffFile'.

jose committed:
Minimized number of functions.

jose committed:
Take into account the granularity level when parsing java files.

jose committed:
Added issue URL.

jose committed:
Pass over analysis and analytics.

jose committed:
Minor update.

jose committed:
Added missing import.

jose committed:
'repr' is the implemented one.

jose committed:
Added __str__ to Component in case any code needs a str representation.

jose committed:
Fixed classes in classes, or methods in methods, classes in methods, ...

jose committed:
Look for children if there are any.

jose committed:
Ignoring 'UnicodeDecodeError' such as "'utf-8' codec can't decode byte 0xe9".

jose committed:
Strip path.

jose committed:
Updated some test cases (work in progress).

jose committed:
Removed extra space.

jose committed:
Added Jose's fork of the pure python java parser and tools javalang (https://github.com/c2nes/javalang) as git submodule, which supports start/end position of blocks of code.

jose committed:
Adapted extraction of start_line/end_line number of each component now that javalang supports it.

jose committed:
Added lt operator so that set of elements can be sorted.

jose committed:
Fixed get_classes method.

jose committed:
Fixed get_methods/get_functions method.

jose committed:
Updated comment.

jose committed:
Fixed version_b.

jose committed:
Position might be None.

jose committed:
Abstract common functions.

jose committed:
A 'file', for example, might not have a start_line/end_line.

jose committed:
Updated comments.

jose committed:
A renamed and untouched file have all same components but name is different.

jose committed:
Fixed 'rename' operation.

jose committed:
Updated learner given the refactor that has been made.

jose committed:
Updated tests.

jose committed:
Remove any "

jose committed:
Replaced exception.

jose committed:
Start/End line might be None.

jose committed:
Get *all* components within a line range.

jose committed:
Updated javalang submodule.

jose committed:
Install javalang submodule.

jose committed:
Fix install submodule.

jose committed:
A line only has one line number.

jose committed:
Minor optimization: if a parent component is not in range, no sub-component could be in range.

jose committed:
Ignore java files for which JavaLang's throws a LexerError.

andrefreitas commented:
Update: I'm lacking the time and energy to follow up on this. To not create false expectation, i'm closing this issue. 

It's better to run your own fork of Schwa and adapt it to your use cases.

Thanks for your time @jose and all the best.
